### PR TITLE
#378 - add special check for empty holes of lists

### PR DESCRIPTION
**Pull request by vicchig**

### CODE REVIEW

File: src/syntax-tree/ast.ts [hunk 1]
```diff
@@ -2577,6 +2577,25 @@ export class ListLiteralExpression extends Expression {
         this.performTypeUpdatesOnInsertInto(insertCode);
     }
 
+    isHolePlacementValid(): boolean {
+        const emptyHolePlacements = this.getEmptyHolesWIndex();
+        return emptyHolePlacements.length === 0
+            ? true
+            : emptyHolePlacements.length === 1 && emptyHolePlacements[0][1] === this.tokens.length - 2;
```
Comment (vicchig) [inline src/syntax-tree/ast.ts:2584]: Only the last hole in the list can be empty.

File: src/syntax-tree/module.ts [hunk 2]
```diff
@@ -544,7 +544,7 @@ export class Module {
         while (Q.length > 0) {
             let curr: CodeConstruct = Q.splice(0, 1)[0];
 
-            if (curr instanceof TypedEmptyExpr) {
+            if (curr instanceof TypedEmptyExpr && !curr.isListElement()) {
```
Comment (vicchig) [inline src/syntax-tree/module.ts:547]: List holes are handled by lists in a lower case.

File: src/syntax-tree/module.ts [hunk 3]
```diff
@@ -563,7 +563,15 @@ export class Module {
                     this.addHighlightToConstruct(curr, ERROR_HIGHLIGHT_COLOUR);
                 }
             } else if (curr instanceof Expression && curr.tokens.length > 0) {
-                Q.push(...curr.tokens);
+                const addHighlight = curr instanceof ListLiteralExpression && !curr.isHolePlacementValid();
+                ret = addHighlight ? CodeStatus.ContainsEmptyHoles : ret;
+
+                for (let i = 0; i < curr.tokens.length; i++) {
+                    if (curr.tokens[i] instanceof TypedEmptyExpr && addHighlight && i < curr.tokens.length - 2) {
```
Comment (vicchig) [inline src/syntax-tree/module.ts:570]: If the hole placement is not valid all holes, but the last one should be highlighted.

File: src/syntax-tree/ast.ts (hunk 1 above)
Comment (MajeedKazemi) [reply src/syntax-tree/ast.ts:2584]: excatly!

**Pull request merged**